service-base/src/main/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementHttpEndpoint.java

```diff
@@ -0,0 +1,242 @@
+    private static final String SPAN_NAME_UPDATE_DEVICE = "update Device from management API";
+    private static final String SPAN_NAME_REMOVE_DEVICE = "remove Device from management API";
+
+    /**
+     * Creates an endpoint for a Vertx instance.
+     *
+     * @param vertx The Vertx instance to use.
+     * @throws NullPointerException if vertx is {@code null};
+     */
+    @Autowired
+    public AbstractDeviceManagementHttpEndpoint(final Vertx vertx) {
+        super(vertx);
+    }
+
+    @Override
+    protected String getEventBusAddress() {
+        return "";
+    }
+
+    @Override
+    public String getName() {
+        return String.format("%s/%s",
+                RegistryManagementConstants.API_VERSION,
+                RegistryManagementConstants.DEVICES_HTTP_ENDPOINT);
+    }
+
+    @Override
+    public void addRoutes(final Router router) {
+
+        final String pathWithTenant = String.format("/%s/:%s", getName(), PARAM_TENANT_ID);
+        final String pathWithTenantAndDeviceId = String.format("/%s/:%s/:%s", getName(), PARAM_TENANT_ID,
+                PARAM_DEVICE_ID);
+
+        // Add CORS handler
+        router.route(pathWithTenant).handler(createCorsHandler(config.getCorsAllowedOrigin(), EnumSet.of(HttpMethod.POST)));
+        router.route(pathWithTenantAndDeviceId).handler(createDefaultCorsHandler(config.getCorsAllowedOrigin()));
+
+
+        // CREATE device with auto-generated deviceID
+        router.post(pathWithTenant)
+                .handler(this::extractOptionalJsonPayload)
+                .handler(this::doCreateDevice);
+
+        // CREATE device
+        router.post(pathWithTenantAndDeviceId)
+                .handler(this::extractOptionalJsonPayload)
+                .handler(this::doCreateDevice);
+
+        // GET device
+        router.get(pathWithTenantAndDeviceId)
+                .handler(this::doGetDevice);
+
+        // UPDATE existing device
+        router.put(pathWithTenantAndDeviceId)
+                .handler(this::extractRequiredJsonPayload)
+                .handler(this::extractIfMatchVersionParam)
+                .handler(this::doUpdateDevice);
+
+        // DELETE device
+        router.delete(pathWithTenantAndDeviceId)
+                .handler(this::extractIfMatchVersionParam)
+                .handler(this::doDeleteDevice);
+    }
+
+    /**
+     * The service to forward requests to.
+     *
+     * @return The service to bind to, must never return {@code null}.
+     */
+    protected abstract DeviceManagementService getService();
+
+    private void doGetDevice(final RoutingContext ctx) {
+
+        final String deviceId = getRequestParam(PARAM_DEVICE_ID, ctx);
+        final String tenantId = getRequestParam(PARAM_TENANT_ID, ctx);
+
+        final HttpServerResponse response = ctx.response();
+
+        final Span span = Util.newChildSpan(SPAN_NAME_GET_DEVICE, TracingHandler.serverSpanContext(ctx), tracer, tenantId, deviceId, getClass().getSimpleName());
```

Comment:
FMPOV it would be desirable to start the span *before* extracting the mandatory parameters and then log any missing parameters to the span. That way, it is easier for support people to determine why a request failed.